### ISSUE TITLE
Overhaul of moderation/blocking tools needed

### ISSUE THREAD

**Issue by ThibG** [tootsuite/mastodon#8186]
I think there is a need for an overhaul of the moderation and user-side blocking tools, making them more flexible and more consistent.

On the admin side of things, we currently can (correct me if I'm wrong):
1. Silence an account, which hides their toots from the public timelines and prevent them from getting in the notifications of people who do not follow them. Still allows them to see toots of any user.
2. Block accounts, which makes them unable to interact with anyone on the instance, or read any of the toots (unless using the public pages of the instance)
3. Silence instances, which has the same effect as silencing all users of that instance
4. Block instances, which has the same effect as blocking all users of that instance (and do not deliver anything to that instance, so following/follower desyncs have no consequence)
5. Optionally, reject/delete media files from an instance

I think this is pretty good, although I would argue that muting someone (or an instance) from public timelines and preventing them from getting in the notifications of people who do not follow them should be two different settings and the same could be said about instances (see #7153).

The terminology could also be changed to be less confusing, or alternatively, get rid of that terminology entirely by using checkboxes: an user/instance could be “blocked” or a combination of “hidden from public TLs” and “blocked unless followed” (see also #5694).

For end-users, now, things get a bit more confusing and inconsistent with “user domain blocks”: they roughly correspond to blocking a whole instance with the exception that you can still follow users there.
Like with admin instance silencing, hiding from public timelines and hiding notifications should be two different settings (see #7792). Furthermore, force-removing followers should be optional.

In addition to all these changes, whether a block is visible to the blocked user should be considered more thoroughly: indeed, I can see the point of not making the blocked user aware, but it is difficult or even impossible to achieve, and leads to more confusing software behaviors and make things harder to debug, especially in a federated setting.

To sum up, I would like to make at least these changes:
- Split “hide from timelines” and “hide from notifications (unless followed)” for both user and instance silences (admin)
- Split “domain hiding” (user) into domain block (removes any kind of relationship, rejects notifications and follow requests), and domain silence (with “hide from timelines” and “hide from notifications unless followed” as two independent options) with also the option to unfollow everyone on that instance (and maybe notifying the remote instance in the process, unlike what is currently done)

**Comment by deutrino** [comment]
Related bug: #9028

**Comment by ThisIsMissEm** [comment]
I'll have an article out soon expanding more on this, but in my FIRES paper (which the mastodon team have access to), I proposed taking a step further, and having a "federation firewall": essentially merging the `domain_allows` and `domain_blocks` tables, and defining a default federation policy. (new table would be like `domain_federation_policies`)
